### ISSUE TITLE
Permissions error when trying to update settings on 3.23.x

### ISSUE THREAD

**Issue by daniellockyer** [TryGhost/Ghost#12026]
reports:

* https://forum.ghost.org/t/saving-settings-configs-error-permission-error-cannot-edit-setting-attempted-to-access-core-setting-from-external-request/15731
* https://forum.ghost.org/t/upgrade-3-22-issues/15667/11

### Issue Summary

Updating to 3.23.x from 3.21.1, and then trying to edit settings will result in the following error:

![https://aws1.discourse-cdn.com/business4/uploads/ghost2/original/2X/9/9f5f5244d0e6f19d57a374fb99d1432f2dda42f7.png](https://aws1.discourse-cdn.com/business4/uploads/ghost2/original/2X/9/9f5f5244d0e6f19d57a374fb99d1432f2dda42f7.png)

### To Reproduce

1. `ghost install 3.21.1 --local`
1. `ghost upgrade 3.23.0 --local`
1. Run the following query:

```
sqlite> select `key`,`type`,`group` from settings where key like 'portal_%';
portal_name|boolean|portal
portal_button|boolean|portal
portal_plans|array|portal
portal_button_style|string|core
portal_button_icon|string|core
portal_button_signup_text|string|core
```

* `portal_button_*` settings seem to have been placed in the `core` group, instead of the desired `portal` group.
* `core` settings cannot be set from outside Ghost, hence the permissions error.

### Technical details:

* Ghost Version: 3.23.x
* Node Version: any
* Browser/OS: any
* Database: SQLite (probably any)

---

Tasks

- [ ] We must to add explicit group and flag migrations for any new setting.
- [ ] We should add a new test to ensure this happens. The quick version for this would be an integrity check for default-settings.json - as this at least stops edits to that file without thinking.

**Comment by rishabhgrg** [comment]
The root cause for this behavior is related to change in settings table and init flow in 3.22.

Any new setting added since version 3.23 (and future versions) , when migrating from any < 3.22 version will always end up with `group` column value as `core` due to this issue. Reason ->

1. The before hook of knex-migrate [here](https://github.com/TryGhost/Ghost/blob/master/core/server/data/migrations/hooks/migrate/before.js#L8)  populates all default settings from default-settings.json, which is called as soon as DB backup is done and before running through migrations.

2. `populateDefaults` adds all the new settings added since old version we are migrating from, and logic [here](https://github.com/TryGhost/Ghost/blob/master/core/server/models/settings.js#L243-L245) ensures any missing columns in db schema are ignored. This means all  the new settings are added to DB with only their `type` value, while the `group` or `flag` values are skipped since the columns don’t exist yet.

3. Migrations in 3.22 run, which adds [here](https://github.com/TryGhost/Ghost/blob/master/core/server/data/migrations/versions/3.22/03-add-group-and-flags-to-settings.js) new group  column to db defaulting to value core for all settings.

4. Another migration in 3.22 [here](https://github.com/TryGhost/Ghost/blob/master/core/server/data/migrations/versions/3.22/04-populate-settings-groups-and-flags.js) updates all the known settings till 3.22 to correct group, but that means all the new settings added after 3.22 stay with group as `core` in DB

5. Since we don’t allow editing `core` settings from external resource, the new settings when attempted to save from Admin throw the permission error

**Comment by sbounmy** [comment]
Same issue here from 3.14.0 -> 3.23.0

**Comment by rishabhgrg** [comment]
Note: We are working on releasing a fix for this in next release asap, meanwhile please avoid migrating to any `3.23.x` version unless you are on version `3.22.x` already.

**Comment by ErisDS** [comment]
Fundamentally the problem here is a 🐔  and 🥚  situation between our prerequisite to populateDefaults before boot vs our need to change the table structure for settings during migrations.

For now the solution is:

- We must to add explicit group and flag migrations for any new setting.
- We should add a new test to ensure this happens. The quick version for this would be an integrity check for default-settings.json - as this at least stops edits to that file without thinking.
- We should figure out a way to resolve this long term E.g. with a major change in 4.0 that allows us to stop doing this.

I have added the first two as tasks to this issue and added the requirement to fix this in 4.0 to our 4.0 tracker.

**Comment by rishabhgrg** [comment]
[3.24.0](https://github.com/TryGhost/Ghost/releases/tag/3.24.0) is now released with the fix.
